KAFKA-6718 / Add rack awareness configurations to StreamsConfig

lkokhreidze opened the pull request:
This PR is part of [KIP-708](https://cwiki.apache.org/confluence/display/KAFKA/KIP-708%3A+Rack+awareness+for+Kafka+Streams) and adds rack aware standby task assignment logic.

Rack aware standby task assignment won't be functional until all parts of this KIP gets merged.

Splitting PRs into three smaller PRs to make the review process easier to follow. Overall plan is the following:

⏭️  Rack aware standby task assignment logic https://github.com/apache/kafka/pull/10851
⏭️  Protocol change, add clientTags to SubscriptionInfoData https://github.com/apache/kafka/pull/10802
👉  Add required configurations to StreamsConfig (public API change, at this point we should have full functionality)

This PR implements last point of the above mentioned plan.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

lkokhreidze committed:
KAFKA-6718 / Add client.tag. prefix constant in StreamsConfig

lkokhreidze committed:
Merge branch 'trunk' into KAFKA-6718-streamsconfig-changes

cadonna commented:
@lkokhreidze Is this PR ready for review?

lkokhreidze commented:
Hi @cadonna 
Not yet. Will open it for review by Monday.

lkokhreidze committed:
KAFKA-6718 / Update logging and finalize standby task assignment logic

lkokhreidze commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -144,10 +148,15 @@
     public static final int DUMMY_THREAD_INDEX = 1;
     public static final long MAX_TASK_IDLE_MS_DISABLED = -1;
 
+    public static final int MAX_RACK_AWARE_ASSIGNMENT_TAG_LIST_SIZE = 10;
+    public static final int MAX_RACK_AWARE_ASSIGNMENT_TAG_KEY_LENGTH = 50;
+    public static final int MAX_RACK_AWARE_ASSIGNMENT_TAG_VALUE_LENGTH = 100;
```
Do not have any science behind those numbers, but thought it's better to have some restrictions.

showuon replied:
Actually, I don't think we need this limitation. Could we not be able to support list size > 10, key len > 50, value len > 100? I guess we can. If so, why this limitation? And like what you've asked, why this number?

lkokhreidze replied:
> Actually, I don't think we need this limitation. Could we not be able to support list size > 10, key len > 50, value len > 100?

I think we need some limitations because this information will be encoded into the subscription info and subscription info is the part of the group metadata message that we need to write into the internal topic. This was also discussed in the voting thread.

As for why those numbers, I really can't say. It can be more, can be less. But we need some limitation I think. But I can't imagine valid use case of having more than 10 tags with more than 50 and 100 characters length.

showuon replied:
Make sense to me. Could we add a comment above these line, to mention we have these limitation because these tag value will be encoded into subscription info, sth like that? Thanks.

cadonna replied:
I am a bit concerned about the length. I would propose the following:

```
MAX_RACK_AWARE_ASSIGNMENT_TAG_LIST_SIZE = 5
MAX_RACK_AWARE_ASSIGNMENT_TAG_KEY_LENGTH = 20
MAX_RACK_AWARE_ASSIGNMENT_TAG_VALUE_LENGTH = 30
```

We can always increase these numbers in future if they are too small, but if we want to decrease them, we might run into backwards compatibility issues.
 
But more important, we should have an integration test that tests the maximum allowed size.

lkokhreidze replied:
Thanks for the comment. I will add the integration test to test all three of those configurations.

showuon replied:
@cadonna , Good suggestion!

lkokhreidze commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -1146,6 +1187,36 @@ private void verifyEOSTransactionTimeoutCompatibility() {
         }
 
+        // validate rack awareness related configuration
+        final List<String> rackAwareAssignmentTags = getList(RACK_AWARE_ASSIGNMENT_TAGS_CONFIG);
```
Not sure if this is the best place to perform those validations, but couldn't find any better place.

showuon replied:
Yes, put validation here is good.

showuon replied:
Maybe put the rack awareness validation into an another method. (If we decided to have these limitation)

lkokhreidze committed:
KAFKA-6718 / Change log level.

lkokhreidze commented on streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:
```diff
@@ -0,0 +1,340 @@
+import static org.apache.kafka.streams.integration.utils.IntegrationTestUtils.safeUniqueTestName;
+
+@Category({IntegrationTest.class})
+public class RackAwarenessIntegrationTest {
```
Since core logic tests are already written as unit tests, this integration tests verifies basic high level functionality.

lkokhreidze commented:
Ready for the review @cadonna @showuon

lkokhreidze committed:
KAFKA-6718 / Remove unused constant

lkokhreidze commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java:
```diff
@@ -142,15 +142,6 @@ private void assignStandbyReplicaTasks(final TreeMap<UUID, ClientState> clientSt
     }
 
-    // Visible for testing
-    static StandbyTaskAssignor createStandbyTaskAssignor(final AssignmentConfigs configs) {
```
Refactored based on the previous PR comment https://github.com/apache/kafka/pull/10851#discussion_r817650626

lkokhreidze commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java:
```diff
@@ -49,7 +53,11 @@ static int pollClientAndMaybeAssignAndUpdateRemainingStandbyTasks(final Map<UUID
         }
 
-        return numRemainingStandbys;
+        log.warn("Unable to assign {} of {} standby tasks for task [{}]. " +
```
Refactored based on previous PR comment
https://github.com/apache/kafka/pull/10851#discussion_r817629787

Since both standby task assignors have the same logging, moved the logging to this utility method.

showuon replied:
This doesn't look right. We log the warning to user even when `numRemainingStandbys==0`?

lkokhreidze replied:
Stupid mistake. Fixed. Also added tests.

lkokhreidze commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java:
```diff
@@ -211,17 +194,32 @@ static void assignStandbyTasksToClientsWithDifferentTags(final ConstrainedPriori
         if (numRemainingStandbys > 0) {
             pendingStandbyTasksToClientId.put(activeTaskId, activeTaskClient);
             tasksToRemainingStandbys.put(activeTaskId, numRemainingStandbys);
+            log.warn("Rack aware standby task assignment was not able to assign {} of {} standby tasks for the " +
```
Added logging based on the previous PR comment https://github.com/apache/kafka/pull/10851#discussion_r817629787

lkokhreidze commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java:
```diff
@@ -211,17 +194,32 @@ static void assignStandbyTasksToClientsWithDifferentTags(final ConstrainedPriori
             countOfUsedClients++;
             numRemainingStandbys--;
 
+            log.debug("Assigning {} out of {} standby tasks for an active task [{}] with client tags {}. " +
```
This was not requested during the previous PR review, but I think it can be useful for debugging to have extra logs.

cadonna commented on clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:
```diff
@@ -1121,6 +1121,27 @@ public String toString() {
+            this.maxSize = maxSize;
+        }
+
+        public static ListSize max(final int maxSize) {
```
Could you rename this method to somthing a bit more specific like `atMostOfLength()`?

cadonna commented on clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java:
```diff
@@ -722,4 +723,24 @@ public void testNiceTimeUnits() {
+                               Importance.HIGH,
+                               "lst doc");
+    }
+
```
Could you add a unit test, so that there is one with a list that is longer than the maximum, one with a list that is strictly shorter than maximum, and one with list that has maximum length?

lkokhreidze commented:
There are few tests failing. Will be fixing them now.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:
```diff
@@ -0,0 +1,340 @@
+        createAndStart(topology, buildClientTags("eu-central-1a", "k8s-cluster-3"), asList(TAG_ZONE, TAG_CLUSTER), numberOfStandbyReplicas);
+        createAndStart(topology, buildClientTags("eu-central-1b", "k8s-cluster-3"), asList(TAG_ZONE, TAG_CLUSTER), numberOfStandbyReplicas);
+        createAndStart(topology, buildClientTags("eu-central-1c", "k8s-cluster-3"), asList(TAG_ZONE, TAG_CLUSTER), numberOfStandbyReplicas);
+
```
I think, here it makes sense to wait until all Streams clients are `RUNNING` so that we know that the rebalance is done.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:
```diff
@@ -0,0 +1,340 @@
+        createAndStart(topology, buildClientTags("eu-central-1b", "k8s-cluster-1"), asList(TAG_ZONE, TAG_CLUSTER), numberOfStandbyReplicas);
+        createAndStart(topology, buildClientTags("eu-central-1b", "k8s-cluster-1"), asList(TAG_ZONE, TAG_CLUSTER), numberOfStandbyReplicas);
+        createAndStart(topology, buildClientTags("eu-central-1b", "k8s-cluster-1"), asList(TAG_ZONE, TAG_CLUSTER), numberOfStandbyReplicas);
+
```
See my comment below about waiting for `RUNNING`.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:
```diff
@@ -0,0 +1,340 @@
+
+        waitUntilRackAwareTaskDistributionIsReached(TAG_ZONE, TAG_CLUSTER);
+    }
+
```
Could you add a test where standbys cannot be distributed over all tags to check that task assignment works in that case?

cadonna reviewed (commented):
@lkokhreidze Thanks for the PR!

Here my comments.

lkokhreidze committed:
KAFKA-6718 / Fix and add tests

lkokhreidze committed:
KAFKA-6718 / Address CR comments

lkokhreidze commented:
Thanks @cadonna for the feedback.
I've addressed your comments.

lkokhreidze committed:
KAFKA-6718 / Rename atMostOfLength -> atMostOfSize

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -1464,6 +1464,10 @@ protected byte uniqueField() {
         return uniqueField;
     }
 
+    protected Map<String, String> clientTags() {
+        return clientTags;
```
This seems to be only used in its own test.

cadonna replied:
Sorry, my bad. I misinterpreted the code.

lkokhreidze replied:
It is indeed only used in tests. But seems like that's a common pattern in the `StreamsPartitionAssignor`. Followed the same approach. Happy to remove this if you think it's unnecessary.

lkokhreidze committed:
KAFKA-6718 / Fix tests yet again

lkokhreidze committed:
Merge branch 'trunk' into KAFKA-6718-streamsconfig-changes

showuon commented on streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:
```diff
@@ -0,0 +1,392 @@
+        waitUntilPartialTaskDistributionIsReachedForTags(singletonList(TAG_CLUSTER));
+    }
+
+    private void waitUntilAllKafkaStreamsClientsAreRunning() throws Exception {
+        TestUtils.waitForCondition(
+            () -> kafkaStreamsInstances.stream().allMatch(it -> KafkaStreams.State.RUNNING == it.kafkaStreams.state()),
+            IntegrationTestUtils.DEFAULT_TIMEOUT,
+            String.format("Kafka Streams did not transit to state %s in %d milliseconds",
+                          KafkaStreams.State.RUNNING, IntegrationTestUtils.DEFAULT_TIMEOUT)
+        );
+    }
```
We already have `startApplicationAndWaitUntilRunning` or `waitForApplicationState` in `IntegrationTestUtils`. I think we can use `startApplicationAndWaitUntilRunning` in our tests.

showuon commented on streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:
```diff
@@ -0,0 +1,392 @@
+
+        waitUntilAllKafkaStreamsClientsAreRunning();
+        waitUntilIdealTaskDistributionIsReachedForTags(singletonList(TAG_ZONE));
+        waitUntilPartialTaskDistributionIsReachedForTags(singletonList(TAG_CLUSTER));
```
Since we've confirmed the stream clients are in running state, should we also `wait` for the tasks distribution? I think we can directly verify them, right?

lkokhreidze replied:
Yes good point. Will update the test.

showuon reviewed (commented):
@lkokhreidze , thanks for the PR. Left some comments.

lkokhreidze committed:
KAFKA-6718 / CR

lkokhreidze commented:
Thanks for the review @showuon 
I've addressed your comments.

showuon commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java:
```diff
@@ -26,17 +27,20 @@
 final class StandbyTaskAssignmentUtils {
-    private StandbyTaskAssignmentUtils() {}
+    private StandbyTaskAssignmentUtils() {
+    }
```
nit: `{}`

showuon reviewed (commented):
Thanks for the update. Left some more comments.

showuon commented on clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java:
```diff
@@ -722,4 +724,35 @@ public void testNiceTimeUnits() {
+    @Test
+    public void testNoExceptionIsThrownWhenListSizeEqualsTheLimit() {
+        final List<String> lst = asList("a", "b", "c");
+        new ConfigDef().define("lst",
```
We can use `assertDoesNotThrow` here and below.

lkokhreidze replied:
Done

showuon commented on clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:
```diff
@@ -1121,6 +1121,27 @@ public String toString() {
+            @SuppressWarnings("unchecked")
+            List<String> values = (List<String>) value;
+            if (values.size() > maxSize) {
+                throw new ConfigException(name, value, "exceeds maximum list size of [" + maxSize + "].");
```
I'm not quite sure if the `value` will output the list strings. Should we use `values` instead?

lkokhreidze replied:
It should output list of strings. Added test to verify.

showuon reviewed (commented):
Made another pass. Left a comment. Thanks.

lkokhreidze committed:
KAFKA-6718 / CR

lkokhreidze commented:
Hi @showuon 
Thanks for taking a look. I've addressed your comments.

cadonna reviewed (commented):
@lkokhreidze Thank you for the updates!

I have one major comment.

lkokhreidze committed:
KAFKA-6718 / CR

lkokhreidze commented:
Thanks @cadonna for the review.
I've addressed your comment in this commit https://github.com/apache/kafka/pull/11837/commits/fbdb17016a3674e234f6ecdbe672472ee405f52d

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:
```diff
@@ -108,55 +117,122 @@ public void cleanup() throws IOException {
     }
 
+    @Test
+    public void shouldThrowConfigExceptionWhenRackAwareAssignmentTagsExceedTheLimit() {
```
I think we misunderstood each other. I was asking for an integration test that verifies that the rebalancing works for the maximum length of the tags. To see if the subscription has still a size that works. Maybe also in combination with many tasks.

lkokhreidze replied:
Apologies, my mistake.
Makes sense. I've added `RackAwarenessIntegrationTest#shouldDoRebalancingWithMaximumNumberOfClientTags` that verifies rebalancing, as well as that rack awareness works after rebalancing too.

lkokhreidze replied:
> Maybe also in combination with many tasks.

FYI - current topology generates 6 tasks, I guess that should be enough?

cadonna replied:
I was more thinking about 1000 stateful tasks, because the task offset sums might also inflate the subscription.

lkokhreidze replied:
Was able to push up to a two hundred stateful tasks and even that was already slow. After 500, test was getting flaky on my computer and sometimes it was just failing due to timeouts.

Feels like these kind of test is too heavy for a local integration tests. Any other way we could do it? Or any pointers to any of the existing integration tests that do the similar thing?

Anyway, I've pushed the changes. I also found one bug in the `ClientTagAwareStandbyTaskAssignor` which is now fixed.

cadonna replied:
I see. I asked some other people and apparently the size of the subscription is not that much of a concern currently. I would still propose to keep the test with 100 instead of 200 tasks. With 200 tasks, the test takes ~10 s on my laptop whereas with 100 tasks it takes ~6 s. I think 6 s is acceptable and might also reduce flakiness. @lkokhreidze and @showuon WDYT?

lkokhreidze replied:
Sounds good to me. Will push changes with 100 stateful tasks for the rebalancing test.

lkokhreidze replied:
Pushed changes to reduce number of stateful tasks to 100
https://github.com/apache/kafka/pull/11837/commits/e6f717229acb4850aec1f33e13cd88cdc3ccb4bb

showuon replied:
+1 for 100 tasks. Thanks.

lkokhreidze replied:
Even with 100 tasks tests on CI failed with timeout. I've bumped timeout to 1.5 minutes in the latest commit.

cadonna replied:
If after your PR the tests are still flaky, let's just test with 6 tasks and merge this PR. You can then add a system test for the scenario with larger subscriptions in a follow-up PR.

lkokhreidze replied:
Thanks @cadonna 
If no objections, I will do that now. The fact that it times out in 1 minute doesn't give much confidence that it won't be flaky.
I will follow up the system test.
fyi @showuon

lkokhreidze committed:
KAFKA-6718 / CR

lkokhreidze committed:
KAFKA-6718 / Add extra validation to ensure that generated client tags have maximum allowed size.

lkokhreidze committed:
KAFKA-6718 / Improve rebalancing integration test; fix ClientTagAwareStandbyTaskAssignor bug

showuon commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java:
```diff
@@ -60,7 +60,7 @@ public boolean assign(final Map<UUID, ClientState> clients,
         final Map<TaskId, Integer> tasksToRemainingStandbys = computeTasksToRemainingStandbys(
             numStandbyReplicas,
-            allTaskIds
+            statefulTaskIds
```
nice catch, and thanks for adding a test for it.

showuon commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtilsTest.java:
```diff
@@ -55,38 +65,38 @@ public void setup() {
 
         assertTrue(ACTIVE_TASKS.stream().allMatch(activeTask -> tasksToRemainingStandbys.get(activeTask) == 0));
         assertTrue(areStandbyTasksPresentForAllActiveTasks(1));
+        verifyNoInteractions(logMock);
```
Thanks for the tests

showuon commented on streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:
```diff
@@ -0,0 +1,438 @@
+    @Test
+    public void shouldDoRebalancingWithMaximumNumberOfClientTags() throws Exception {
+        initTopology(100, 2);
+        final Duration timeout = Duration.ofMinutes(3);
```
3 mins for a test is too long IMO. Could we reduce to 2 mins or less?

showuon reviewed (commented):
@lkokhreidze , thanks for the update. Overall LGTM! Only 1 concern about the new added test, and its flakiness. Please help fix it. Thanks.

https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11837/11/
```
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
```

lkokhreidze committed:
KAFKA-6718 / Reduce test duration

lkokhreidze commented:
Thanks @showuon for the feedback. Should be better now.
Also reduced test timeout to 1.5 minutes.

cadonna reviewed (approved):
Thank you @lkokhreidze !

LGTM!

lkokhreidze committed:
KAFKA-6718 / Reduce number of stateful tasks to 100

lkokhreidze committed:
KAFKA-6718 / Bump timeout to 1.5 minutes to reduce flakiness

showuon commented:
@lkokhreidze , let's make sure the tests stable before we merge this final piece of PRs. Thanks.

lkokhreidze committed:
KAFKA-6718 / Reduce number of stateful tasks to 9

lkokhreidze commented:
@showuon 
I've reduced number of stateful tasks to 9. This should eliminate the flakiness. 
I will do the system test in the follow-up PR.

cadonna commented:
@lkokhreidze We also need some updates to the docs. The sections to update are https://kafka.apache.org/31/documentation/streams/architecture#streams_architecture_recovery and maybe also https://kafka.apache.org/31/documentation/streams/developer-guide/config-streams.html.

lkokhreidze commented:
Thanks @cadonna ,
I assume it's okay to do it in a follow-up PR?

cadonna commented:
Yes, doing it in a follow-up PR is totally fine!

showuon reviewed (approved):
LGTM! As long as the tests are stable, we can merge it. Thanks.

cadonna commented:
Merged!
@lkokhreidze Congrats to this milestone! Nice work! 🥳

lkokhreidze commented:
Big thank you @cadonna and @showuon for all the reviews.

showuon commented:
Nice work @lkokhreidze ! And nice team work, @cadonna ! :)
